src/ucp/wireup/select.c

```diff
@@ -1520,6 +1520,22 @@ static void ucp_wireup_init_keepalive_map(ucp_worker_h worker,
 
     dev_map_used = 0;
 
+    /* exclude from ep_check map all devices where built-in
+     * keepalive is supported - mark all such devices as used */
+    for (lane = 0; lane < key->num_lanes; ++lane) {
```

Comment:
can we unite the 2 loops?

Reply:
no: we need 2 loops (nested or pipe-lined) in any case
@brminich wanted to propose any unification, but still no news from him

Reply:
seems to be no clear way to unite given generic use-case.

Reply:
right, I didn't find any was... I can add some macro to make code bit shorter, but in fact there will be same pair of loops

Reply:
how about
```
for (lane ...) {
    ...
    dev_index = ...
    iface_attr = ...
    if (FLAG_EP_KEEPALIVE) {
        dev_map_used |= UCS_BIT(dev_index);
    } else if ((!(UCS_BIT(dev_index) & dev_map_used) && EP_CHECK) {
        kep_ep_check_map |= ...
        dev_map_used |= UCS_BIT(dev_index);
    }
}
```

Reply:
or maybe precompute it during worker init?

Reply:
> how about

will not work: in case if any lane-bit is added to ep_check map and later we found that another lane on same device has built-in keepalive, then we have to remove such lane-bit from ep_check map - this is issue. and we can't pre-compile it in worker init because in general case there is no correllation between ep_check_map and devices

Reply:
ok. so please add comments.
to 1st loop: /* find all devices with built-in keepalive support */
for 2nd loop: /* send ep_check on devices without built-in keepalive */

Reply:
updated comments